flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/BroadcastRecordWriter.java

```diff
@@ -84,7 +86,7 @@ void randomEmit(T record, int targetChannelIndex) throws IOException, Interrupte
 		if (bufferBuilder != null) {
 			for (int index = 0; index < numberOfChannels; index++) {
 				if (index != targetChannelIndex) {
-					addBufferConsumer(bufferBuilder.createBufferConsumer(), index);
+					addBufferConsumer(randomTriggeredConsumer.copyAndSync(bufferBuilder), index);
```

Comment:
Uhhh. This is a nice spaghetti (the way how `randomTriggeredConsumer` is being set and used here).

I don't know, but maybe we should re-think the inheritance and code deduplication between `BroadcastRecordWriter` and `RecordWriter`? Maybe we should drop this connection and live with code duplication? Or maybe there is some better way? Like extract some class and switch to composition?

If you and @curcur are fine with the current fix, I would be fine with doing the refactor/rethinking it outside of this bug fix.

Reply:
I agree, but without the refactoring, I don't think we can avoid spaghetti.

I'd suggest to do a refactoring as a separate task.